packages/lib/src/vis/tiles/TiledLayer.tsx

```diff
@@ -1,51 +1,61 @@
-import { sum } from 'lodash';
-import { Suspense } from 'react';
+import { Suspense, useRef } from 'react';
 import { LinearFilter, NearestFilter, Vector2 } from 'three';
+import type { Group } from 'three';
 
 import { useCameraState } from '../hooks';
+import type { Size } from '../models';
 import { useAxisSystemContext } from '../shared/AxisSystemProvider';
 import Tile from './Tile';
 import type { TilesApi } from './api';
 import type { ColorMapProps } from './models';
 import {
   getTileOffsets,
-  getScaledVisibleDomains,
+  getScaledVisibleBox,
   sortTilesByDistanceTo,
 } from './utils';
 
 interface Props extends ColorMapProps {
   api: TilesApi;
   layer: number;
+  size: Size;
 }
 
 function TiledLayer(props: Props) {
-  const { api, layer, ...colorMapProps } = props;
+  const { api, layer, size: meshSize, ...colorMapProps } = props;
 
   const { baseLayerIndex, numLayers, tileSize } = api;
   const layerSize = api.layerSizes[layer];
 
-  const { abscissaConfig, ordinateConfig, visSize } = useAxisSystemContext();
-  const { xVisibleDomain, yVisibleDomain } = useCameraState(
-    (...args) => getScaledVisibleDomains(...args, layerSize),
-    [layerSize]
+  const groupRef = useRef<Group>(null);
+  const { abscissaConfig, ordinateConfig } = useAxisSystemContext();
+  const box = useCameraState(
+    (...args) => getScaledVisibleBox(...args, meshSize, layerSize, groupRef),
+    [meshSize, layerSize, groupRef]
   );
 
```

Comment:
If `useCameraState` returns the group's visible box (unscaled) like I suggested in another comment, am I wrong in thinking that the box of the group rendered by `TiledHeatmapMesh` would always be the same as the box of the groups rendered by the layers? If so, then maybe the visible box can be passed to `TiledLayer` as a prop and this component would then just have to call `getScaledVisibleBox` (or `scaleBoxToLayer` as per my other comment) with the right `layerSize`? Apart from the obvious performance improvement of one fewer `useCameraState`, it would save you from dealing with an undefined `box` below.